_posts/2020-09-07-introduction-to-system-tests.md

```diff
@@ -0,0 +1,275 @@
+of class contains the external port name, external port 9094 with protocol TCP and more.
+Everything what this type of test validates is only single behaviour of a sofware module that we defined
+previously.
+We don't need a running `Kubernetes cluster` for unit tests.
+It is only in-memory representation of class, in this case the `Kafka` model.
+
+```java
+@Test
+public void testExternalNodePorts() {
+    Kafka kafkaAssembly = new KafkaBuilder(ResourceUtils.createKafka(namespace, cluster, replicas,
+            image, healthDelay, healthTimeout, metricsCm, configuration, emptyMap()))
+            .editSpec()
+                .editKafka()
+                    .withNewListeners()
+                        .withNewKafkaListenerExternalNodePort()
+                            .withNewKafkaListenerAuthenticationTlsAuth()
+                            .endKafkaListenerAuthenticationTlsAuth()
+                        .endKafkaListenerExternalNodePort()
+                    .endListeners()
+                .endKafka()
+            .endSpec()
+            .build();
+    KafkaCluster kc = KafkaCluster.fromCrd(kafkaAssembly, VERSIONS);
+
+    // Check StatefulSet changes
+    StatefulSet sts = kc.generateStatefulSet(true, null, null);
+
+    List<ContainerPort> ports = sts.getSpec().getTemplate().getSpec().getContainers().get(0).getPorts();
+    assertThat(ports.contains(kc.createContainerPort(KafkaCluster.EXTERNAL_PORT_NAME, KafkaCluster.EXTERNAL_PORT, "TCP")), is(true));
+
+    // Check external bootstrap service
+    Service ext = kc.generateExternalBootstrapService();
+    assertThat(ext.getMetadata().getName(), is(KafkaCluster.externalBootstrapServiceName(cluster)));
+    assertThat(ext.getSpec().getType(), is("NodePort"));
+    assertThat(ext.getSpec().getSelector(), is(kc.getSelectorLabels().toMap()));
+    assertThat(ext.getSpec().getPorts(), is(Collections.singletonList(kc.createServicePort(KafkaCluster.EXTERNAL_PORT_NAME, KafkaCluster.EXTERNAL_PORT, KafkaCluster.EXTERNAL_PORT, "TCP"))));
+    checkOwnerReference(kc.createOwnerReference(), ext);
+
+    // Check per pod services
+    for (int i = 0; i < replicas; i++)  {
+        Service srv = kc.generateExternalService(i);
+        assertThat(srv.getMetadata().getName(), is(KafkaCluster.externalServiceName(cluster, i)));
+        assertThat(srv.getSpec().getType(), is("NodePort"));
+        assertThat(srv.getSpec().getSelector().get(Labels.KUBERNETES_STATEFULSET_POD_LABEL), is(KafkaCluster.kafkaPodName(cluster, i)));
+        assertThat(srv.getSpec().getPorts(), is(Collections.singletonList(kc.createServicePort(KafkaCluster.EXTERNAL_PORT_NAME, KafkaCluster.EXTERNAL_PORT, KafkaCluster.EXTERNAL_PORT, "TCP"))));
+        checkOwnerReference(kc.createOwnerReference(), srv);
+    }
+}
+```
+
+### Integration tests
+
+Next up are the integration tests, which move one level higher.
+You can find all these tests inside our `api` module.
```

Comment:
By this definition the `KafkaAssemblyOperatorTest` and `KafkaAssemblyOperatorMockTest` (etc.) are also integration tests of a sort, and probably a better example than anything from the `api` module. For me the distinction has always been a bit blurry.

There's also the `TopicOperatorIntegrationTest` which runs against a real Kafka cluster and a real Kubernetes, though the only things in that Kube cluster are `KafkaTopic`s.